disklabel/disklabel.8
.\" Copyright (c) 1987, 1988 Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Symmetric Computer Systems.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	from: @(#)disklabel.8	6.8 (Berkeley) 6/24/90
.\"	$Id: disklabel.8,v 1.5 1993/08/05 20:49:16 jtc Exp $
.\"
.Dd June 24, 1990
.Dt DISKLABEL 8
.Os
.Sh NAME
.Nm disklabel
.Nd read and write disk pack label
.Sh SYNOPSIS
.Nm disklabel
.Op Fl r
.Ar disk
.Nm disklabel
.Fl w
.Op Fl r
.Ar disk
.Ar disktype
.Op Ar packid Op Ar priboot secboot
.Nm disklabel
.Fl e
.Op Fl r
.Ar disk
.Nm disklabel
.Fl R
.Op r
.Ar disk
.Ar protofile
.Op Ar priboot secboot
.Nm disklabel
.Op Fl NW
.Op disk
.Sh DESCRIPTION
.Nm Disklabel
can be used to install, examine or modify the label on a disk drive or pack.
When writing the label, it can be used
to change the drive identification,
the disk partitions on the drive,
or to replace a damaged label or bootstrap.
The disk label is located on one of the first sectors
of each disk (usually block 0).
On machines that require a block-0 bootstrap (VAX 11/750),
the label is inserted into the bootstrap program.
This information is used by the system disk driver and by the bootstrap
program to determine how to program the drive.
There are several forms of the command that display, install or modify
the label on a disk.
Each form has an additional option,
.Fl r ,
which causes the label to be read from or written to the disk directly,
rather than going through the system's in-core copy of the label.
When writing, the in-core copy is also updated if possible.
This option may allow a label to be installed on a disk
without kernel support for a label, such as when labels are first installed
on a system; it must be used when first installing a label on a disk.
.Pp
The first form of the command is used to examine the label on the named
disk drive (e.g. hp0 or /dev/rhp0c).
It will display all of the parameters associated with the drive
and its partition layout.
Unless the
.Fl r
flag is given,
the kernel's in-core copy of the label is displayed;
if the disk has no label, or the partition types on the disk are incorrect,
the kernel may have constructed or modified the label.
If the
.Fl r
flag is given, the label from the raw disk will be displayed rather
than the in-core label.
.Pp
The second form of the command, with the
.Fl w
flag, is used to write a standard label on the designated drive.
The required arguments to
.Nm disklabel
are the drive to be labelled (e.g. hp0), and
the drive type as described in the
.Xr disktab 5
file.
The drive parameters and partitions are taken from that file.
If different disks of the same physical type are to have different
partitions, it will be necessary to have separate disktab entries
describing each, or to edit the label after installation as described below.
The first optional argument is a pack identification string,
up to 16 characters long.
The pack id must be quoted if it contains blanks.
If the
.Fl r
flag is given, the disk sectors containing the label and bootstrap
will be written directly,
otherwise the existing label will be updated in place without modifying
the bootstrap.
If the disk does not already have a label, the
.Fl r
flag must be used.
In either case, the kernel's in-core label is replaced.
.Pp
Alternate versions of the bootstrap files may be specified
after the pack identifier.
If an alternate bootstrap is not specified, the standard bootstrap will be used.
The bootstrap programs are located in
.Pa /usr/mdec .
The names of the bootstrap programs may be specified in
.Xr disktab  5 ;
if not specified,
the default names are of the form
.Em basename No boot
for the primary (block 0) bootstrap, and
.No boot Em basename
for the secondary (block 1-15) bootstrap;
for example,
.Pa /usr/mdec/hpboot
and
.Pa /usr/mdec/boothp
if the disk device is
.Em hp0 .
.Pp
An existing disk label may be edited by using the
.Fl e
flag.
The label is read from the in-core kernel copy,
or directly from the disk if the
.Fl r
flag is also given.
The label is formatted and then supplied to an editor for changes.
If no editor is specified in an EDITOR environment variable,
.Xr vi 1
is used.
When the editor terminates, the formatted label is reread
and used to rewrite the disk label.
.Pp
With the
.Fl R
flag,
.Nm disklabel
is capable of restoring a disk label that was formatted
in a prior operation and saved in an ascii file.
The prototype file used to create the label should be in the same format
as that produced when reading or editing a label.
Comments are delimited by
.Em #
and newline.
If the
.Fl r
option is also given,
a block-0 bootstrap is installed on machines that use one;
either the disktype or the names of the bootstrap files must be specified
on such machines.
.Pp
Finally, the
.Fl NW
flags for
.Nm disklabel
explicitly disallow and allow, respectively, writing of the pack label area
on the selected disk.
.Sh FILES
.Bl -item -compact
.It
.Pa /etc/disktab
.It
.Pa /usr/mdec/ Ns Sy xx Ns Pa boot
.It
.Pa /usr/mdec/boot Ns Sy xx
.Sh SEE ALSO
.Xr disktab 5 ,
.Xr disklabel 5
.Sh DIAGNOSTICS
The kernel device drivers will not allow the size of a disk partition
to be decreased or the offset of a partition to be changed while it is open.
Some device drivers create a label containing only a single large partition
if a disk is unlabeled; thus, the label must be written to the
.Dq a
partition of the disk while it is open.
This sometimes requires the desired label to be set in two steps,
the first one creating at least one other partition,
and the second setting the label on the new partition
while shrinking the
.Dq a
partition.
.Sh BUGS
When a disk name is given without a full pathname,
the constructed device name uses the
.Dq c
partition on the vax and the
.Dq a
partition on the tahoe.


doc/release_notes
1.6 Release [ Jan 19, 2018 ]  changes over 1.5beta:
	Added preliminary support of sectors != 512 bytes
	Added -I flag for edumpfs to invert meaning of displayed bits
	Improved invalid FS handling for newfs and fsck
	Improved failure code display for eln command
	dbprint will now exit cleanly if the debug port is already open
	Fixed directory corrupting bug in file dirent create code
	fsck command now checks for sufficient stack at startup
	Switched compiler from DICE 2.06.34 to DICE 3.16
	Added support for AS225 "ls" ACTION_EX_OBJECT / ACTION_EX_NEXT
	Added ls code to show UNIX block/char device major/minor numbers
	Added ls code to display full date and time (with -T option)
	Changed ls sorted filename display to be per column instead of per row
	The ls utility can now use passwd and group files for symbolic names
	Added unmount command
	Fixed sign bug in block free search which confused the frag allocator 
	    (file writes) when it knew a specific block must be available
	Improved block free search for 1:8 ratio frag:block filesystems
	Changed BFFSTool "PreAlloc" field display to be hexadecimal
	Relabeled filesystem now retains the name in the superblock
	Fixed a corruption bug caused by metadata flush while still in use
	Amiga attributes Pure, Script, and Delete are now preserved in inode
	Fixed bug where root dir contents were being reported as root dirs
	Lock on "" and "." should now work the same
	Fixed bug in BSD44 label search code for command line tools
	Added port of BSD 4.4 filesystem dump utility

1.5beta Release  [ Aug 21, 1996 ]  changes over 1.4beta0:
	Fixed problem where fs allowed deletion of locked file
	Fixed low disk space condition where file shortened incorrectly
		if fs runs out of space during write (two bugs).
	Made many changes to allow 386i FFS filesystems to be read
	Fixed inode comments for sockets and fifos
	Fixed Addbuffers problem - corruption on removing buffers
	Dumpfs (filesystem info) now supports little-endian filesystems
	Removed potential memory leak in symlink resolution
	Added rest of support for ReadLink packet.
	Changed BFFSTool bytes read/written display to Kilobytes
	Centralized configuration variables to single file
	File permission checks added for file and directory read/write
	Renamed dumpfs edumpfs and added automatic x86 fs interpreting
	Added new port of 4-Nov-94 BSD dumpfs, newfs, fsck, tunefs, diskpart
	Added new packets for UNIX chown and utimes support
	Added mknod command
	Wrote touch which can set all three UNIX times, System V feel
	Added support for BSD4.4 fastlinks
	Fixed write bug where fsize=bsize and 2nd level inode indirect needed
	Performance enhancement - allocation algorithm change to stop
		cg write thrashing on large files
	Added port of 1-Oct-95 BSD restore
	Added chown command
	Added chmod command - supports UNIX ### and ugo+=perms methods, and -R
	Enhanced ls command to feel more like UNIX ls (sorting, recursive, etc)

1.4beta0 Release  [ Nov 28, 1994 ]  changes over 1.31:
	Misc Enforcer/Mungwall hits fixed in filesystem, fsck, tool, others
		now that I have a machine that can capture those problems.
	Slight performance enhancement made for sequential (multiblock) reads
	Fixed fsck problem with sometimes recognizing the superblock.
	Changed make options such that executable is residentable (can be
		multiply mounted at boot), this solves problem of having
		more than one automounted BFFS partition in RDB area.
	Cleaned up various aspects of the code to ease the transition
		to the opposite endian compatibility compilation.
	Fixed bug where string not terminated properly for symlink comments.
		Thanks to Ty Sarna for pointing out this problem.
	Disk Key now represents the file inode number to be more compatible
		with ixemul.librarary; thanks to Wolfgang Baron for suggestion
	Did some code optimizing to speed packet interpretation (mainly
		changed how packet data is passed between routines)
	Fixed AmigaDOS mode path parsing problem.  Thanks to Wolfgang Baron
		for demonstrating this bug.
	Fixed fsck, newfs bug when given a name which is not a disk device
	Added support for ACTION_GET_DISK_FSSM packet, as created by
		Ralph Babel and suggested by Christopher A. Wichura
	Fixed inode corruption problem which occurred during writes over
		files (not creating new ones) when filesystem is full
	Problem of BFFS crash with handlers not supporting disk change
		packet eliminated, thanks to Michael Hitch for bug report
	Fixed file write bug where inode data could get flushed from the
		cache before write is done using that data.
	Converted packet type lookups to table for those that are possible,
		will speed up packet service
	Rewrote file fragment group reader for greater efficiency / speed; will
		use item in cache if available and doesn't flush dirty frags
	Fixed examine of volume where root directory size was incorrectly
		returned - thanks to Andrew H. McCardie for pointing this out`
	Fixed problem of resolving assigned names with mixed case.  Thanks
		to Russell McOrmond for bug report.

1.31 Release  [ 18 Mar 1994 ]  Beta Release changes over 1.3:
	Unknown packets are now reported as ERROR_ACTION_NOT_KNOWN instead
		of ERROR_NOT_IMPLEMENTED.  Apparently AmigaDOS and other
		programs look only for ERROR_ACTION_NOT_KNOWN.  Thanks to
		for pointing this out.
	Problem with unaligned buffer file reads where the number of
		requested fragments is greater than cache size fixed.
	Low memory (bytes 0-3) trashing stopped.  Thanks to reports from
		Joe Vasher and Antti Miettinen
	ACTION_EXAMINE_FH fixed, thanks to Joe Vasher for pointing out problem
	Partitions with fragment sizes other than 1k are now mountable
	512 byte block fragments now supported in directory routines
	Problem with allocation where last cylinder group is not full
		has been corrected - would result in filesystem corruption

1.3 Release  [ 2 Feb 1994 ]  changes over 1.3beta:
	If device does not exist, BFFS no longer GURUs the machine.

unix_support/disklabel.c
/*
 * Copyright (c) 1983, 1987 Regents of the University of California.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)disklabel.c	5.17 (Berkeley) 2/23/91";
#endif /* LIBC_SCCS and not lint */

#define DKTYPENAMES
#include <stdio.h>
#include "sys/param.h"
#include "ufs/fs.h"
#include "sys/disklabel.h"
#include "paths.h"
#include <sys/file.h>
#ifdef cdh
#include <err.h>
#endif

static	char *dgetstr();
static	dgetent();
static	dnamatch();
static	dgetnum();
static	dgetflag();
static	gettype();
static	error();
char bootarea[BBSIZE];

/* next block added by cdh */
#define errno 0
#define EFTYPE 0
extern char bootarea[];
extern int  DEV_BSIZE;
int force_autolabel = 0;

struct disklabel *
getdiskbyname(char *name)
/* const removed by cdh */
{
	static struct	disklabel disk;
	static char	boot[BUFSIZ];
	char	*buf;
	char	*localbuf;
	char	*cp, *cq;	/* can't be register */
	register struct	disklabel *dp = &disk;
	register struct partition *pp;
	char	p, max, psize[3], pbsize[3],
		pfsize[3], poffset[3], ptype[3];
	u_long	*dx;

	buf = (char *) malloc(BUFSIZ);
	if (buf == NULL) {
		fprintf(stderr, "fatal: unable to allocate memory\n");
		exit(2);
	}

	if (dgetent(buf, name) <= 0) {
		free(buf);
		return ((struct disklabel *)0);
	}

	localbuf = (char *) malloc(BUFSIZ);
	if (localbuf == NULL) {
		fprintf(stderr, "fatal: unable to allocate memory\n");
		exit(2);
	}

	bzero((char *)&disk, sizeof(disk));
	/*
	 * typename
	 */
	cq = dp->d_typename;
	cp = buf;
	while (cq < dp->d_typename + sizeof(dp->d_typename) - 1 &&
	    (*cq = *cp) && *cq != '|' && *cq != ':')
		cq++, cp++;
	*cq = '\0';
	/*
	 * boot name (optional)  xxboot, bootxx
	 */
	cp = boot;
	dp->d_boot0 = dgetstr("b0", &cp);
	dp->d_boot1 = dgetstr("b1", &cp);
	cp = localbuf;
	cq = dgetstr("ty", &cp);
	if (cq && strcmp(cq, "removable") == 0)
		dp->d_flags |= D_REMOVABLE;
	else  if (cq && strcmp(cq, "simulated") == 0)
		dp->d_flags |= D_RAMDISK;
	if (dgetflag("sf"))
		dp->d_flags |= D_BADSECT;

#define getnumdflt(field, dname, dflt) \
	{ int f = dgetnum(dname); \
	(field) = f == -1 ? (dflt) : f; }

	getnumdflt(dp->d_secsize, "se", DEV_BSIZE);
	dp->d_ntracks = dgetnum("nt");
	dp->d_nsectors = dgetnum("ns");
	dp->d_ncylinders = dgetnum("nc");
	cq = dgetstr("dt", &cp);
	if (cq)
		dp->d_type = gettype(cq, dktypenames);
	else
		getnumdflt(dp->d_type, "dt", 0);
	getnumdflt(dp->d_secpercyl, "sc", dp->d_nsectors * dp->d_ntracks);
	getnumdflt(dp->d_secperunit, "su", dp->d_secpercyl * dp->d_ncylinders);
	getnumdflt(dp->d_rpm, "rm", 3600);
	getnumdflt(dp->d_interleave, "il", 1);
	getnumdflt(dp->d_trackskew, "sk", 0);
	getnumdflt(dp->d_cylskew, "cs", 0);
	getnumdflt(dp->d_headswitch, "hs", 0);
	getnumdflt(dp->d_trkseek, "ts", 0);
	getnumdflt(dp->d_bbsize, "bs", BBSIZE);
	getnumdflt(dp->d_sbsize, "sb", SBSIZE);
	strcpy(psize, "px");
	strcpy(pbsize, "bx");
	strcpy(pfsize, "fx");
	strcpy(poffset, "ox");
	strcpy(ptype, "tx");
	max = 'a' - 1;
	pp = &dp->d_partitions[0];
	for (p = 'a'; p < 'a' + MAXPARTITIONS; p++, pp++) {
		psize[1] = pbsize[1] = pfsize[1] = poffset[1] = ptype[1] = p;
		pp->p_size = dgetnum(psize);
		if (pp->p_size == -1)
			pp->p_size = 0;
		else {
			pp->p_offset = dgetnum(poffset);
			getnumdflt(pp->p_fsize, pfsize, 0);
			if (pp->p_fsize)
				pp->p_frag = dgetnum(pbsize) / pp->p_fsize;
			getnumdflt(pp->p_fstype, ptype, 0);
			if (pp->p_fstype == 0 && (cq = dgetstr(ptype, &cp)))
				pp->p_fstype = gettype(cq, fstypenames);
			max = p;
		}
	}
	dp->d_npartitions = max + 1 - 'a';
	(void)strcpy(psize, "dx");
	dx = dp->d_drivedata;
	for (p = '0'; p < '0' + NDDATA; p++, dx++) {
		psize[1] = p;
		getnumdflt(*dx, psize, 0);
	}
	dp->d_magic = DISKMAGIC;
	dp->d_magic2 = DISKMAGIC;
	free(buf);
	free(localbuf);
	return (dp);
}

#include <ctype.h>

static	char *tbuf;
static	char *dskip();
static	char *ddecode();

/*
 * Get an entry for disk name in buffer bp,
 * from the diskcap file.  Parse is very rudimentary;
 * we just notice escaped newlines.
 */
static
dgetent(bp, name)
	char *bp, *name;
{
	register char *cp;
	register int c;
	register int i = 0, cnt = 0;
	char *ibuf;
	int tf;

	tbuf = bp;
	tf = open(_PATH_DISKTAB, 0);
	if (tf < 0) {
		perror(_PATH_DISKTAB);
/*		error(errno); */
		return (-1);
	}

	ibuf = (char *) malloc(BUFSIZ);
	if (ibuf == NULL) {
		fprintf(stderr, "fatal: unable to allocate memory\n");
		exit(2);
	}

	for (;;) {
		cp = bp;
		for (;;) {
			if (i == cnt) {
				cnt = read(tf, ibuf, BUFSIZ); /* disktab file */
				if (cnt <= 0) {
					perror("read");
/*					error(errno); */
					close(tf);
					free(ibuf);
					return (0);
				}
				i = 0;
			}
			c = ibuf[i++];
			if (c == '\n') {
				if (cp > bp && cp[-1] == '\\'){
					cp--;
					continue;
				}
				break;
			}
			if (cp >= bp+BUFSIZ) {
				error(EFTYPE);
				break;
			} else
				*cp++ = c;
		}
		*cp = 0;

		/*
		 * The real work for the match.
		 */
		if (dnamatch(name)) {
			close(tf);
			free(ibuf);
			return (1);
		}
	}
	free(ibuf);
}

/*
 * Dnamatch deals with name matching.  The first field of the disktab
 * entry is a sequence of names separated by |'s, so we compare
 * against each such name.  The normal : terminator after the last
 * name (before the first field) stops us.
 */
static
dnamatch(np)
	char *np;
{
	register char *Np, *Bp;

	Bp = tbuf;
	if (*Bp == '#')
		return (0);
	for (;;) {
		for (Np = np; *Np && *Bp == *Np; Bp++, Np++)
			continue;
		if (*Np == 0 && (*Bp == '|' || *Bp == ':' || *Bp == 0))
			return (1);
		while (*Bp && *Bp != ':' && *Bp != '|')
			Bp++;
		if (*Bp == 0 || *Bp == ':')
			return (0);
		Bp++;
	}
}

/*
 * Skip to the next field.  Notice that this is very dumb, not
 * knowing about \: escapes or any such.  If necessary, :'s can be put
 * into the diskcap file in octal.
 */
static char *
dskip(bp)
	register char *bp;
{

	while (*bp && *bp != ':')
		bp++;
	if (*bp == ':')
		bp++;
	return (bp);
}

/*
 * Return the (numeric) option id.
 * Numeric options look like
 *	li#80
 * i.e. the option string is separated from the numeric value by
 * a # character.  If the option is not found we return -1.
 * Note that we handle octal numbers beginning with 0.
 */
static
dgetnum(id)
	char *id;
{
	register int i, base;
	register char *bp = tbuf;

	for (;;) {
		bp = dskip(bp);
		if (*bp == 0)
			return (-1);
		if (*bp++ != id[0] || *bp == 0 || *bp++ != id[1])
			continue;
		if (*bp == '@')
			return (-1);
		if (*bp != '#')
			continue;
		bp++;
		base = 10;
		if (*bp == '0')
			base = 8;
		i = 0;
		while (isdigit(*bp))
			i *= base, i += *bp++ - '0';
		return (i);
	}
}

/*
 * Handle a flag option.
 * Flag options are given "naked", i.e. followed by a : or the end
 * of the buffer.  Return 1 if we find the option, or 0 if it is
 * not given.
 */
static
dgetflag(id)
	char *id;
{
	register char *bp = tbuf;

	for (;;) {
		bp = dskip(bp);
		if (!*bp)
			return (0);
		if (*bp++ == id[0] && *bp != 0 && *bp++ == id[1]) {
			if (!*bp || *bp == ':')
				return (1);
			else if (*bp == '@')
				return (0);
		}
	}
}

/*
 * Get a string valued option.
 * These are given as
 *	cl=^Z
 * Much decoding is done on the strings, and the strings are
 * placed in area, which is a ref parameter which is updated.
 * No checking on area overflow.
 */
static char *
dgetstr(id, area)
	char *id, **area;
{
	register char *bp = tbuf;

	for (;;) {
		bp = dskip(bp);
		if (!*bp)
			return (0);
		if (*bp++ != id[0] || *bp == 0 || *bp++ != id[1])
			continue;
		if (*bp == '@')
			return (0);
		if (*bp != '=')
			continue;
		bp++;
		return (ddecode(bp, area));
	}
}

/*
 * Tdecode does the grung work to decode the
 * string capability escapes.
 */
static char *
ddecode(str, area)
	register char *str;
	char **area;
{
	register char *cp;
	register int c;
	register char *dp;
	int i;

	cp = *area;
	while ((c = *str++) && c != ':') {
		switch (c) {

		case '^':
			c = *str++ & 037;
			break;

		case '\\':
			dp = "E\033^^\\\\::n\nr\rt\tb\bf\f";
			c = *str++;
nextc:
			if (*dp++ == c) {
				c = *dp++;
				break;
			}
			dp++;
			if (*dp)
				goto nextc;
			if (isdigit(c)) {
				c -= '0', i = 2;
				do
					c <<= 3, c |= *str++ - '0';
				while (--i && isdigit(*str));
			}
			break;
		}
		*cp++ = c;
	}
	*cp++ = 0;
	str = *area;
	*area = cp;
	return (str);
}

#define strcasecmp stricmp
static
gettype(t, names)
	char *t;
	char **names;
{
	register char **nm;

	for (nm = names; *nm; nm++)
		if (strcasecmp(t, *nm) == 0)
			return (nm - names);
	if (isdigit(*t))
		return (atoi(t));
	return (0);
}

static
error(err)
	int err;
{
	char *p;

	(void)fprintf(stderr, "disklabel: ", 12);
	(void)fprintf(stderr, _PATH_DISKTAB, sizeof(_PATH_DISKTAB) - 1);
	p = strerror(err);
	(void)fprintf(stderr, p, strlen(p));
	(void)fprintf(stderr, "\n", 1);
}

/*
 * Fetch disklabel for disk.
 */
int readlabel(lpp)
	struct disklabel *lpp;
{
	struct disklabel *lp;

	if (bread(bootarea, 0, BBSIZE))
		goto autosize_partition;

	if (force_autolabel)
	    goto autosize_partition;

	/* Search for disk label */
	for (lp = (struct disklabel *)bootarea;
	    lp <= (struct disklabel *)(bootarea + BBSIZE - sizeof(*lp));
	    lp = (struct disklabel *)((char *)lp + 16)) {
		if (lp->d_magic == DISKMAGIC &&
		    lp->d_magic2 == DISKMAGIC) break;
	}
	if (lp > (struct disklabel *)(bootarea + BBSIZE - sizeof(*lp)) ||
	    lp->d_magic != DISKMAGIC || lp->d_magic2 != DISKMAGIC)
	    goto autosize_partition;

	bcopy(lp, lpp, sizeof(struct disklabel));
	return(0);

	autosize_partition:
	force_autolabel = 0;
#ifdef AMIGA
	if (dio_label(lpp))
		if (file_label(lpp))
			return(1);
	return(0);
#else
	if (file_label(lpp))
		return(1);
	return(0);
#endif
}

writelabel(boot, lp)
	char *boot;
	register struct disklabel *lp;
{
#ifndef cdh
	register int i;
	int flag;
#endif

	lp->d_magic = DISKMAGIC;
	lp->d_magic2 = DISKMAGIC;
	lp->d_checksum = 0;

	if (bwrite(boot, 0, lp->d_bbsize)) {
		perror("write");
		return (1);
	}
	return (0);
}
